### ISSUE TITLE
Better utilize GPU resources for PGD with random restarts

### ISSUE THREAD

**Issue by SohamTamba** [MadryLab/robustness#98]
The current implementation for PGD with random restarts works like this:

**Input: X, Y, model**
```
output = X.clone()
for restart_i in range(num_random_restarts):
    noise = random_constrained_noise(shape=X.shape) if restart_i > 0 else 0
    pert_X = X + noise
    adv, is_misclass = run_pgd(pert_X, Y, model)
   output[is_misclass] = adv[is_misclass] 
return output
```
If the user has enough GPU memory - which will often be the case for CIFAR10 -, then a better way to go about this is:
```
B, C, H, W = X.shape
X_stack = torch.stack([X for _ in range(num_random_restarts)])
Y_stack = torch.stack([Y for _ in range(num_random_restarts)])

noise = random_constrained_noise(shape=[num_random_restarts, B, C, H, W])
noise[0, :, :, :, :] = 0
pert_X_stack = X_stack + noise

pert_X_stack = pert_X_stack.view(-1, C, H, W)
Y_stack = Y_stack.view(-1)

adv, is_misclass = run_pgd(pert_X_stack, Y_stack, model)
adv = adv.view(-1, B, C, H, W)
is_misclass = is_misclass.view(-1, B)
return_ind = is_misclass.argmax(axis=0) 
return adv[return_ind]
```

My only concern is this might negatively affect BatchNorm. But given how computationally intensive adversarial training is, this might be a worthwhile option to provide users.

**Let me know if you guys are interested in a PR.**

Reference: [](https://github.com/MadryLab/robustness/blob/79d371fd799885ea5fe5553c2b749f41de1a2c4e/robustness/attacker.py#L235)
